// ng-angular-popup/esm2022/lib/ng-toast.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NgToastComponent } from './ng-toast.component';
import * as i0 from "@angular/core";
class NgToastModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.1", ngImport: i0, type: NgToastModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.1.1", ngImport: i0, type: NgToastModule, declarations: [NgToastComponent], imports: [CommonModule], exports: [NgToastComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.1.1", ngImport: i0, type: NgToastModule, imports: [CommonModule] }); }
}
export { NgToastModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.1", ngImport: i0, type: NgToastModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgToastComponent
                    ],
                    imports: [
                        CommonModule,
                    ],
                    exports: [
                        NgToastComponent,
                    ]
                }]
        }] });
